File: br/pkg/lightning/backend/sharedisk/reader.go
// Copyright 2023 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package sharedisk

import (
	"context"
	"encoding/binary"
	"io"
	"time"

	"github.com/pingcap/log"
	"github.com/pingcap/tidb/br/pkg/storage"
	"github.com/pingcap/tidb/metrics"
	"github.com/pingcap/tidb/util/logutil"
	"github.com/pingcap/tidb/util/mathutil"
	"go.uber.org/zap"
)

type kvReader struct {
	byteReader *byteReader
	key        []byte
	val        []byte
}

func newKVReader(ctx context.Context, name string, store storage.ExternalStorage, initFileOffset uint64, bufSize int) (*kvReader, error) {
	br, err := newByteReader(ctx, store, name, initFileOffset, bufSize)
	if err != nil {
		return nil, err
	}
	return &kvReader{
		byteReader: br,
		key:        nil,
		val:        nil,
	}, nil
}

func (r *kvReader) nextKV() (key, val []byte, err error) {
	if r.byteReader.eof() {
		return nil, nil, nil
	}
	r.byteReader.reset()
	lenBuf, err := r.byteReader.sliceNext(8)
	if err != nil {
		return nil, nil, err
	}
	keyLen := int(binary.BigEndian.Uint64(lenBuf.get()))
	kBuf, err := r.byteReader.sliceNext(keyLen)
	if err != nil {
		return nil, nil, err
	}
	lenBuf, err = r.byteReader.sliceNext(8)
	if err != nil {
		return nil, nil, err
	}
	valLen := int(binary.BigEndian.Uint64(lenBuf.get()))
	valBuf, err := r.byteReader.sliceNext(valLen)
	if err != nil {
		return nil, nil, err
	}
	return kBuf.get(), valBuf.get(), nil
}

func (r *kvReader) Close() error {
	return r.byteReader.Close()
}

type statsReader struct {
	byteReader *byteReader
	propBytes  []byte
}

func newStatsReader(ctx context.Context, store storage.ExternalStorage, name string, bufSize int) (*statsReader, error) {
	br, err := newByteReader(ctx, store, name, 0, bufSize)
	if err != nil {
		return nil, err
	}
	return &statsReader{
		byteReader: br,
		propBytes:  nil,
	}, nil
}

func (r *statsReader) nextProp() (*RangeProperty, error) {
	if r.byteReader.eof() {
		return nil, nil
	}
	r.byteReader.reset()
	lenBuf, err := r.byteReader.sliceNext(4)
	if err != nil {
		return nil, err
	}
	propLen := int(binary.BigEndian.Uint32(lenBuf.get()))
	if cap(r.propBytes) < propLen {
		r.propBytes = make([]byte, propLen)
	}
	propBytes, err := r.byteReader.sliceNext(propLen)
	if err != nil {
		return nil, err
	}
	return decodeProp(propBytes.get())
}

func (r *statsReader) Close() error {
	return r.byteReader.Close()
}

func decodeProp(data []byte) (*RangeProperty, error) {
	rp := &RangeProperty{}
	keyLen := binary.BigEndian.Uint32(data[0:4])
	rp.Key = data[4 : 4+keyLen]
	rp.Size = binary.BigEndian.Uint64(data[4+keyLen : 12+keyLen])
	rp.Keys = binary.BigEndian.Uint64(data[12+keyLen : 20+keyLen])
	rp.offset = binary.BigEndian.Uint64(data[20+keyLen : 28+keyLen])
	rp.WriterID = int(binary.BigEndian.Uint32(data[28+keyLen : 32+keyLen]))
	rp.DataSeq = int(binary.BigEndian.Uint32(data[32+keyLen : 36+keyLen]))
	return rp, nil
}

type byteReader struct {
	ctx           context.Context
	name          string
	storageReader storage.ReadSeekCloser

	buf       []byte
	bufOffset int

	fileStart uint64

	auxBuf   []byte
	cowSlice [][]byte
	isEOF    bool
}

func newByteReader(ctx context.Context, store storage.ExternalStorage, name string, initFileOffset uint64, bufSize int) (*byteReader, error) {
	storageReader, err := store.Open(ctx, name)
	if err != nil {
		return nil, err
	}
	_, err = storageReader.Seek(int64(initFileOffset), io.SeekStart)
	if err != nil {
		return nil, err
	}
	br := &byteReader{
		ctx:           ctx,
		name:          name,
		storageReader: storageReader,
		buf:           make([]byte, bufSize),
		bufOffset:     bufSize,
	}
	return br, err
}

type slice interface {
	get() []byte
}

type commonSlice []byte

func (c commonSlice) get() []byte {
	return c
}

type lazySlice struct {
	cowIdx     int
	byteReader *byteReader
}

func (s lazySlice) get() []byte {
	return s.byteReader.cowSlice[s.cowIdx]
}

// sliceNext reads the next n bytes from the reader and returns a buffer slice containing those bytes.
// If the reader has fewer than n bytes remaining in current buffer, `auxBuf` is used as a container instead.
func (r *byteReader) sliceNext(n int) (slice, error) {
	b := r.next(n)
	readLen := len(b)
	if readLen == n {
		return mkLazySlice(r, b), nil
	}
	if cap(r.auxBuf) < n {
		r.auxBuf = make([]byte, n)
	}
	r.auxBuf = r.auxBuf[:n]
	copy(r.auxBuf, b)
	for readLen < n {
		r.cloneSlices()
		err := r.reload()
		if err != nil {
			return nil, err
		}
		b = r.next(n - readLen)
		copy(r.auxBuf[readLen:], b)
		readLen += len(b)
	}
	return commonSlice(r.auxBuf), nil
}

func mkLazySlice(r *byteReader, s []byte) lazySlice {
	r.cowSlice = append(r.cowSlice, s)
	return lazySlice{
		cowIdx:     len(r.cowSlice) - 1,
		byteReader: r,
	}
}

func (r *byteReader) reset() {
	r.cowSlice = r.cowSlice[:0]
}

func (r *byteReader) cloneSlices() {
	for i := range r.cowSlice {
		r.cowSlice[i] = append([]byte(nil), r.cowSlice[i]...)
	}
}

func (r *byteReader) eof() bool {
	return r.isEOF && len(r.buf) == r.bufOffset
}

func (r *byteReader) next(n int) []byte {
	end := mathutil.Min(r.bufOffset+n, len(r.buf))
	ret := r.buf[r.bufOffset:end]
	r.bufOffset += len(ret)
	return ret
}

func (r *byteReader) reload() error {
	startTime := time.Now()
	nBytes, err := io.ReadFull(r.storageReader, r.buf[0:])
	if err == io.EOF {
		logutil.BgLogger().Error("unexpected EOF", zap.String("file", r.name), zap.Uint64("start", r.fileStart))
		r.isEOF = true
		return err
	} else if err != nil && err == io.ErrUnexpectedEOF {
		r.isEOF = true
	} else if err != nil {
		logutil.BgLogger().Warn("Other error during reading from external storage", zap.String("file", r.name), zap.Uint64("start", r.fileStart), zap.Error(err))
		return err
	}
	elapsed := time.Since(startTime).Microseconds()
	ReadByteForTest.Add(uint64(nBytes))
	ReadTimeForTest.Add(uint64(elapsed))
	readRate := float64(nBytes) / 1024.0 / 1024.0 / (float64(time.Since(startTime).Microseconds()) / 1000000.0)
	log.Info("s3 read rate", zap.Any("res", readRate))
	metrics.GlobalSortSharedDiskRate.WithLabelValues("read").Observe(readRate)
	ReadIOCnt.Add(1)
	r.bufOffset = 0
	if nBytes < len(r.buf) {
		// The last batch.
		r.buf = r.buf[:nBytes]
	}
	return nil
}

func (r *byteReader) Close() error {
	return r.storageReader.Close()
}

File: rust/crates/tidb-lexer/src/reader.rs
// Copyright 2026 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// See the License for the specific language governing permissions and
// limitations under the License.

//! A byte cursor over the SQL source, mirroring `pkg/parser/lexer.go`'s
//! `reader`. Positions are byte offsets; `peek` past end returns 0, matching
//! the Go reader's EOF sentinel.

#[derive(Debug)]
pub(crate) struct Reader<'a> {
    src: &'a str,
    bytes: &'a [u8],
    off: usize,
}

impl<'a> Reader<'a> {
    pub(crate) fn new(src: &'a str) -> Self {
        Reader {
            src,
            bytes: src.as_bytes(),
            off: 0,
        }
    }

    pub(crate) fn src(&self) -> &'a str {
        self.src
    }

    pub(crate) fn offset(&self) -> usize {
        self.off
    }

    pub(crate) fn set_offset(&mut self, off: usize) {
        self.off = off;
    }

    pub(crate) fn eof(&self) -> bool {
        self.off >= self.bytes.len()
    }

    /// Byte at the current position, or 0 at/after EOF.
    pub(crate) fn peek(&self) -> u8 {
        self.byte_at(self.off)
    }

    /// Byte at an absolute offset, or 0 if out of range.
    pub(crate) fn byte_at(&self, off: usize) -> u8 {
        if off < self.bytes.len() {
            self.bytes[off]
        } else {
            0
        }
    }

    pub(crate) fn inc(&mut self) {
        if self.off < self.bytes.len() {
            self.off += 1;
        }
    }

    pub(crate) fn inc_n(&mut self, n: usize) {
        self.off = (self.off + n).min(self.bytes.len());
    }

    /// Reads the current byte and advances (0 at EOF).
    pub(crate) fn read_byte(&mut self) -> u8 {
        let b = self.peek();
        self.inc();
        b
    }

    /// Advances while `pred` holds on the current byte; returns the byte that
    /// stopped it (0 at EOF), mirroring `reader.incAsLongAs`.
    pub(crate) fn inc_as_long_as<F: Fn(u8) -> bool>(&mut self, pred: F) -> u8 {
        loop {
            let b = self.peek();
            if self.eof() || !pred(b) {
                return b;
            }
            self.inc();
        }
    }

    /// Case-insensitive prefix check at the current position.
    pub(crate) fn starts_with_ci(&self, prefix: &str) -> bool {
        let rest = &self.bytes[self.off.min(self.bytes.len())..];
        let p = prefix.as_bytes();
        if rest.len() < p.len() {
            return false;
        }
        rest[..p.len()].eq_ignore_ascii_case(p)
    }
}
